### PR TITLE
Manually map Safari versions in getMatchingBrowserVersion

### PR DESCRIPTION
**Pull request by foolip**
This is not ideal, but it fixes the problem now. Tests are in place to
ensure that if we fix it in some other way, there's no observable
difference.

### CODE REVIEW

File: scripts/release/mirror.js [hunk 1]
```diff
@@ -12,13 +12,49 @@ const { browsers } = bcd;
...
  */
 
+const matchingSafariVersions = new Map([
+  ['≤4', '≤3'],
+  ['1', '1'],
+  ['1.1', '1'],
+  ['1.2', '1'],
+  ['1.3', '1'],
+  ['2', '1'],
+  ['3', '1'],
```
Comment (gsnedders) [inline scripts/release/mirror.js:22]:
```suggestion
  ['3', '2'],
```

I think? Safari 3 for macOS was WebKit 523.10 and iOS 1.1.3 was WebKit 420.1 and iOS 2.0 525.18.1.

**Review by gsnedders — COMMENTED** [review]
I think the rest of the iOS/macOS mirroring is okay, actually, despite what I said earlier offline. In all other cases, we seem to be okay.

File: scripts/release/mirror.js (hunk 1 above)
Comment (foolip) [reply scripts/release/mirror.js:22]: I've carried this along from https://gist.github.com/foolip/6872cf5350dccf6224fce984fb73fba1 and the UA string in https://github.com/mdn/browser-compat-data/pull/11173#issuecomment-865201634 was:

> Mozilla/5.0 (iPhone; U; CPU iPhone OS 2_0 like Mac OS X; en-us) AppleWebKit/525.18.1 (KHTML, like Gecko) Version/3.1.1 Mobile/5A347 Safari/525.20

How to interpret that "Version/3.1.1" I don't know, but going by the WebKit version you're right.

**Comment by foolip** [comment]
I've compared `build/data.json` before and after these changes and it looks like the only changes are 8&rarr;7 and 15.1&rarr;15.

I don't think the 15.1&rarr;15 changes are worth reviewing at all, since that's just a bug. I'll make a list of the paths with 8&rarr;7 changes.

This will also allow the mirror migration script to replace a lot more data.

**Comment by foolip** [comment]
I hacked up `npm run diff` to produce this:

```
api.Element.matches:
 → support.safari_ios.0.version_added is edited ("8" → "7")
api.HTMLMediaElement.audioTracks:
 → support.safari_ios.version_added is edited ("8" → "7")
api.HTMLMediaElement.videoTracks:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.advance:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.continue:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.delete:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.direction:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.key:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.primaryKey:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.source:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursor.update:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursorWithValue:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBCursorWithValue.value:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.close:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.createObjectStore:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.deleteObjectStore:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.name:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.objectStoreNames:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.transaction:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.version:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBDatabase.versionchange_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBFactory:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBFactory.cmp:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBFactory.deleteDatabase:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBFactory.open:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.count:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.get:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.getKey:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.keyPath:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.multiEntry:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.name:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.objectStore:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.openCursor:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.openKeyCursor:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBIndex.unique:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.bound:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.lower:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.lowerBound:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.lowerOpen:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.only:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.upper:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.upperBound:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBKeyRange.upperOpen:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.add:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.autoIncrement:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.clear:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.count:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.createIndex:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.delete:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.deleteIndex:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.get:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.index:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.indexNames:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.keyPath:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.name:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.openCursor:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.openKeyCursor:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.put:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBObjectStore.transaction:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBOpenDBRequest:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBOpenDBRequest.blocked_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBOpenDBRequest.upgradeneeded_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest.error:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest.error_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest.readyState:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest.result:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest.source:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest.success_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBRequest.transaction:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.abort:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.abort_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.complete_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.db:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.error:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.error_event:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.mode:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBTransaction.objectStore:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBVersionChangeEvent:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBVersionChangeEvent.newVersion:
 → support.safari_ios.version_added is edited ("8" → "7")
api.IDBVersionChangeEvent.oldVersion:
 → support.safari_ios.version_added is edited ("8" → "7")
api.PerformanceNavigationTiming:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.domComplete:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.domContentLoadedEventEnd:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.domContentLoadedEventStart:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.domInteractive:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.loadEventEnd:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.loadEventStart:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.redirectCount:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.toJSON:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.type:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.unloadEventEnd:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.PerformanceNavigationTiming.unloadEventStart:
 → support.safari_ios.version_added is edited ("15.1" → "15")
api.TextTrackList.getTrackById:
 → support.safari_ios.version_added is edited ("8" → "7")
api.indexedDB:
 → support.safari_ios.version_added is edited ("8" → "7")
css.properties.background-repeat.round_space:
 → support.safari_ios.version_added is edited ("8" → "7")
html.elements.source.sizes:
 → support.safari_ios.1.version_added is edited ("8" → "7")
html.elements.source.srcset:
 → support.safari_ios.1.version_added is edited ("8" → "7")
```

**Comment by foolip** [comment]
These 8&rarr;7 changes seemed worth reviewing / testing on iOS 7/8 on BrowserStack:

- [api.HTMLMediaElement.audioTracks/videoTrack](https://mdn-bcd-collector.appspot.com/tests/api/HTMLMediaElement) confirmed in iOS 7 with
- [api.TextTrackList.getTrackById](http://mdn-bcd-collector.appspot.com/tests/api/TextTrackList/getTrackById) was *not* in iOS 7, but was in 8. Probably the desktop data is wrong.
- html.elements.source.sizes/srcset (the content attributes) can't be tested directly, but the data doesn't match the properties on `HTMLSourceElement`. Needs investigation.

Then there's `indexedDB` and `IDB*`. https://caniuse.com/indexeddb

**Comment by foolip** [comment]
https://github.com/mdn/browser-compat-data/pull/16879 updates the IndexedDB data.

**Comment by foolip** [comment]
https://github.com/mdn/browser-compat-data/pull/16880 fixes getTrackById.

File: scripts/release/mirror.js (hunk 1 above)
Comment (gsnedders) [reply scripts/release/mirror.js]: Historically the Safari version string was separately coded for iOS and macOS, and it was likely largely an approximate decision about rough correlations (and there's one or two cases later where the iOS Safari version stayed on the `.0` major version even after trunk WebKit releases!), especially given it was (IIRC) only exposed in the UA string.

I suspect it was decided to call it `3.1.1` as it was closer to that macOS Safari release than previous ones, even though it appears to have branched later than the macOS release.

**Comment by gsnedders** [comment]
> I don't think the 15.1→15 changes are worth reviewing at all, since that's just a bug. 

They're probably worth reviewing?

**Comment by gsnedders** [comment]
> ```
> api.PerformanceNavigationTiming:
>  → support.safari_ios.version_added is edited ("15.1" → "15")
> ```
> 
> I… think this is wrong? I think it should be 15.1 for both macOS and iOS, but I haven't tested them recently.

Ah, no: https://github.com/WebKit/WebKit/commit/988a6bd73c4f41ec442b56e455f3388e9fc2d0b7 is the commit that enabled it, and as the tags indicate that was in Safari 15.0 for macOS, but only Safari 15.1 for iOS.

Turns out I should just trust the changes I made a while back: https://github.com/mdn/browser-compat-data/pull/13142 ;P

**Comment by foolip** [comment]
> > I don't think the 15.1→15 changes are worth reviewing at all, since that's just a bug. 
> 
> 
> 
> They're probably worth reviewing?

My reasoning was that any 15+15.1 pairs that we have are very likely just due to the mirroring script bug. Off-by-one versions further in the past will have been explicitly entered like this and reviewed at some point, making it more likely they're actually correct.

**Comment by foolip** [comment]
@gsnedders sounds like at least one case of 15+15.1 is correct then?

Maybe what we should do is check which such pairs existing before the mirroring migrations.

**Comment by foolip** [comment]
I've run `comm -12 <(npm run traverse -- -b safari -f 15) <(npm run traverse -- -b safari_ios -f 15.1)` at commit f11324f842f52ea54a96de5b0c504145f7b6814f (before da52849b5d5a431bb696265d09954b19f442f329) and these are the APIs that had 15+15.1 pairs:

api.PerformanceNavigationTiming
api.PerformanceNavigationTiming.domComplete
api.PerformanceNavigationTiming.domContentLoadedEventEnd
api.PerformanceNavigationTiming.domContentLoadedEventStart
api.PerformanceNavigationTiming.domInteractive
api.PerformanceNavigationTiming.loadEventEnd
api.PerformanceNavigationTiming.loadEventStart
api.PerformanceNavigationTiming.redirectCount
api.PerformanceNavigationTiming.toJSON
api.PerformanceNavigationTiming.type
api.PerformanceNavigationTiming.unloadEventEnd
api.PerformanceNavigationTiming.unloadEventStart

@gsnedders confirmed these are correct per https://github.com/mdn/browser-compat-data/pull/13142.

**Comment by foolip** [comment]
The remaining changes that will result from this are now:

```
api.Element.matches:
 → support.safari_ios.0.version_added is edited ("8" → "7")
api.HTMLMediaElement.audioTracks:
 → support.safari_ios.version_added is edited ("8" → "7")
api.HTMLMediaElement.videoTracks:
 → support.safari_ios.version_added is edited ("8" → "7")
css.properties.background-repeat.round_space:
 → support.safari_ios.version_added is edited ("8" → "7")
html.elements.source.sizes:
 → support.safari_ios.1.version_added is edited ("8" → "7")
html.elements.source.srcset:
 → support.safari_ios.1.version_added is edited ("8" → "7")
```

**Comment by foolip** [comment]
I've sent https://github.com/mdn/browser-compat-data/pull/16959 to fix the api.Element.matches case.

**Comment by foolip** [comment]
Alright, sizes and srcset fixed in https://github.com/mdn/browser-compat-data/pull/16961, that's the last of it!

**Comment by foolip** [comment]
Alright, the only remaining changes now are correct:

```
api.HTMLMediaElement.audioTracks:
 → support.safari_ios.version_added is edited ("8" → "7")
api.HTMLMediaElement.videoTracks:
 → support.safari_ios.version_added is edited ("8" → "7")
```

**Comment by DaoWays369** [comment]
ThNk

**Review by queengooborg — APPROVED** [review]
I'm still not a fan of the idea of placing this mapping in a script vs. the browser data, but it at least solves the problem right away.  I'll follow up to move this into the browser JSON!

**Pull request merged**